From b3f762ffb4a1ea78da2d79415597ea0434cf28d7 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Tue, 30 Sep 2025 15:09:31 +0200
Subject: [PATCH 408/553] QWinRegistryKey: de-pessimize stringValue()

Don't create a QString object that we might not need (if the optional
already contains one).

Using value_or() with an argument that is not of trivial type is like
having a function default argument of the same type: it's constructed
whether we need it or not; one should use overloading instead
(QTBUG-98117).

Rewrite to the equally-readable check-and-return-else-default-construct.

Amends 40523b68c14bf618bdc2d5438deebf34627be3af.

Pick-to: 6.8
Task-number: QTBUG-98117
Change-Id: Iecd7fa9c3705c8416b5609e6d157079e10bfca59
Reviewed-by: Oliver Wolff <oliver.wolff@qt.io>
(cherry picked from commit 0f3793fea21b17e540265e0f802fdf7b1a6340fe)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/kernel/qwinregistry.cpp | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/corelib/kernel/qwinregistry.cpp b/src/corelib/kernel/qwinregistry.cpp
index fb315cacb7e..37bf3f99ae1 100644
--- a/src/corelib/kernel/qwinregistry.cpp
+++ b/src/corelib/kernel/qwinregistry.cpp
@@ -191,7 +191,9 @@ QVariant QWinRegistryKey::value(const QString &subKey) const
 // Otherwise, the resulting string (which may be empty) is returned.
 QString QWinRegistryKey::stringValue(const wchar_t *subKey) const
 {
-    return value<QString>(subKey).value_or(QString());
+    if (auto v = value<QString>(subKey))
+        return std::move(*v);
+    return QString();
 }
 
 QString QWinRegistryKey::stringValue(const QString &subKey) const
-- 
2.51.2

